### PR TITLE
Dotplot changes draft

### PR DESCRIPTION
**Pull request by ogibson**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1451

#### Link to staging deployment URL 

in UI PR

#### Links to any Pull Requests related to this
UI - [#546](https://github.com/biomage-ltd/ui/pull/546)
API - biomage-ltd/api#253
Worker - biomage-ltd/worker#209

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging
- [ ] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: python/src/worker/tasks/dotplot.py [hunk 1]
```diff
@@ -27,19 +26,35 @@ def _format_result(self, result):
...
-            request["genes"] = self.task_def["genes"]
-        else:
-            request["nGenes"] = self.task_def["nGenes"]
 
-        typeOfSets = self.task_def["typeOfSets"]
+        input = self.task_def["input"]
+        subset = self.task_def["subset"]
 
-        cellSets = get_cell_sets(self.experiment_id)
+        request = {"markerGenes": self.task_def["markerGenes"]}
```
Comment (ogibson) [inline python/src/worker/tasks/dotplot.py]: Change this to a name that makes it clearer that this is a bool

File: python/src/worker/tasks/dotplot.py [hunk 2]
```diff
@@ -27,19 +26,35 @@ def _format_result(self, result):
...
+
+        if self.task_def["markerGenes"]:
+            request["nGenes"] = input["nGenes"]
+        else:
+            request["genes"] = input["genes"]
```
Comment (ogibson) [inline python/src/worker/tasks/dotplot.py]: genes --> listOfGenes

Comment (ogibson) [reply python/src/worker/tasks/dotplot.py]: genes --> geneNames

Comment (kafkasl) [reply python/src/worker/tasks/dotplot.py]: what is the difference between `nGenes` and `genes`?  can't the parameter have the same name?

Comment (aerlaut) [reply python/src/worker/tasks/dotplot.py]: `nGenes` was used to store the number of marker genes to be shown.
`genes` was used to store a list of genes whose expression is requested.

`nGenes` is changed to `numberOfMarkers`, while `genes` is renamed to `customGenesList`.

File: python/src/worker/tasks/dotplot.py (hunk 1 above)
Comment (aerlaut) [reply python/src/worker/tasks/dotplot.py]: this is changed to `useMarkerGenes`

File: python/src/tests/tasks/test_dotplot.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    @pytest.fixture(autouse=True)
+    def load_correct_definition(self):
+        self.correct_request = {
+            "body": {
+                "name": "DotPlot",
+                "subset": {"cellClassKey": "louvain", "cellSetKey": "all"},
+                "markerGenes": True,
+                "input": {"nGenes": 5},
+            }
+        }
```
Comment (ivababukova) [inline python/src/tests/tasks/test_dotplot.py]: this looks like the outdated input

File: r/R/worker_helpers.R [hunk 4]
```diff
@@ -1,3 +1,37 @@
+getTopMarkerGenes <- function(nFeatures, data, cellSets) {
+  data$custom <- NA
+
+  object_ids <- data$cells_id
+  for (i in seq_along(cellSets)) {
+    set <- cellSets[[i]]
+    filtered_cells <- intersect(set$cellIds, object_ids)
+    data$custom[object_ids %in% filtered_cells] <- i
```
Comment (ivababukova) [inline r/R/worker_helpers.R]: Feels like the use of `data$custom` makes the code very fragile. It is being set from `R/dotplot.R`, but it is called `custom` and if anyone decides to rename it, they won't see that it is also going to break this code as well. Is there any way to avoid this and be explicit in the code that this name `custom` is very important, it shouldn't be renamed?
I don't know R very well, but a couple of suggestions that may or may not work well:
- rename "custom" to something more meaningful that shows to the reader that this is an important slot
- pass the results of `data$custom[object_ids %in% filtered_cells]`, the only thing that you use `custom` for in the `getTopMarkerGenes` function directly into the function.
- pass the slot name `custom` as a parameter to the `getTopMarkerGenes` function.
- add an error check that fails the code as early as possible if the slot name in `dotplot.R` doesn't match the name used in `getTopMarkerGenes`.

Basically, be explicit that these are linked, try to avoid situations when they might not

File: r/R/worker_helpers.R [hunk 5]
```diff
@@ -1,3 +1,37 @@
+getTopMarkerGenes <- function(nFeatures, data, cellSets) {
+  data$custom <- NA
```
Comment (ivababukova) [inline r/R/worker_helpers.R]: why are you resetting this after setting it from the `dotplot.R` code?

File: r/R/dotplot.R [hunk 6]
```diff
@@ -1,35 +1,88 @@
...
+    features <- as.data.frame(getMarkerNames(data, all_markers))
+    rownames(features) <- features$input
+  } else {
+    req_genes <- req$body$customGenesList
+    annot <- data@misc$gene_annotations
+    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
+    features <- annot_subset[, c("input", "name")]
+  }
+
+  dotplot_data <- Seurat::DotPlot(data, features = features$input, group.by = "custom")$data
```
Comment (ivababukova) [inline r/R/dotplot.R]: why do you have to do `features = features$input` in here, when in line 72 (`rownames(features) <- features$input`) you've already kind of done that?

File: r/R/dotplot.R [hunk 7]
```diff
@@ -1,35 +1,88 @@
...
+  if (useMarkerGenes) {
+    num_features <- req$body$numberOfMarkers
+    all_markers <- getTopMarkerGenes(num_features, data, group_by_cell_sets)
+    features <- as.data.frame(getMarkerNames(data, all_markers))
+    rownames(features) <- features$input
+  } else {
+    req_genes <- req$body$customGenesList
+    annot <- data@misc$gene_annotations
+    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
+    features <- annot_subset[, c("input", "name")]
```
Comment (ivababukova) [inline r/R/dotplot.R:70]: why in here we don't need to do `rownames(features) <- features$input` the same way we did it on line 72?

File: python/src/worker/tasks/dotplot.py [hunk 8]
```diff
@@ -25,19 +22,29 @@ def _format_result(self, result):
...
 
-        setNames = [set["key"] for set in cellSets]
-        request["cellSets"] = cellSets[setNames.index(typeOfSets)]
+        # Getting the cell ids for subsetting the seurat object with a group of cells.
+        groupByCellSet = [cellSet for cellSet in cellSets if cellSet['key'] == self.task_def["groupBy"]][0]
+        otherCellSets = [cellSet for cellSet in cellSets if cellSet['key'] != self.task_def["groupBy"]]
+
+        filterByDefinition = self.task_def["filterBy"]
+        isFilterByAll = filterByDefinition['group'].lower() == "all"
+        filterByCellSet = otherCellSets
```
Comment (ogibson) [inline python/src/worker/tasks/dotplot.py]: We don't use filterByCellSet, is this line necessary?

File: python/src/worker/tasks/dotplot.py [hunk 9]
```diff
@@ -25,19 +22,29 @@ def _format_result(self, result):
...
+
+        if not isFilterByAll:
+            children = [cellSet for cellSet in cellSets if cellSet["key"] == filterByDefinition['group']][0]["children"]
+            filterByCellSet = [child for child in children if child["key"] == filterByDefinition['key']][0]
```
Comment (ogibson) [inline python/src/worker/tasks/dotplot.py]: We just replace it here.

File: python/src/worker/tasks/dotplot.py [hunk 10]
```diff
@@ -25,19 +22,29 @@ def _format_result(self, result):
...
-        typeOfSets = self.task_def["typeOfSets"]
-
+        # getting cell ids for the groups we want to display.
         cellSets = get_cell_sets(self.experiment_id)
 
-        setNames = [set["key"] for set in cellSets]
-        request["cellSets"] = cellSets[setNames.index(typeOfSets)]
+        # Getting the cell ids for subsetting the seurat object with a group of cells.
+        groupByCellSet = [cellSet for cellSet in cellSets if cellSet['key'] == self.task_def["groupBy"]][0]
+        otherCellSets = [cellSet for cellSet in cellSets if cellSet['key'] != self.task_def["groupBy"]]
```
Comment (ogibson) [inline python/src/worker/tasks/dotplot.py]: We don't use this either

File: r/R/dotplot.R [hunk 11]
```diff
@@ -1,35 +1,84 @@
...
+  }
 
-  res <- purrr::transpose(dotplot_data)
+  dotplot_data <- Seurat::DotPlot(data, features = features$input, group.by = "dotplot.groups")$data
+  # features.plot has the ensemble ids
+  dotplot_data$name <- features[dotplot_data$features.plot, "name"]
+  dotplot_data <- dotplot_data[stringr::str_order(dotplot_data$id, numeric = TRUE), ]
+  dotplot_data <- dotplot_data %>% transmute(cellSets = as.character(id), geneName = as.character(name), avgExpression = avg.exp, cellsPercentage = pct.exp)
```
Comment (alexvpickering) [inline r/R/dotplot.R]: To me it looks like Seurat is using `avg.exp.scaled` column for the dot plot colors/legend. I think this is why ours is not on a similar scale as Seurat and isn't scaled per feature.

File: r/R/dotplot.R [hunk 12]
```diff
@@ -1,35 +1,84 @@
...
+  }
+  # If NA values are left in the group, dotplot function will fail.
+  data <- subset(data, subset = dotplot.groups != "NA")
 
-  dotplot_data <- dotplot_data[, c("avg.exp", "pct.exp", "name", "id")]
-  colnames(dotplot_data) <- c("avgExpression", "cellsPercentage", "geneName", "cellSet")
+  #Get marker genes or requested gene names.
+  if (useMarkerGenes) {
+    num_features <- req$body$numberOfMarkers
+    all_markers <- getTopMarkerGenes(num_features, data, group_by_cell_sets)
```
Comment (alexvpickering) [inline r/R/dotplot.R:63]: I think we want to play around with the parameters to `presto`. Check out [this tutorial](https://divingintogeneticsandgenomics.rbind.io/post/clustered-dotplot-for-single-cell-rnaseq/). Here is the dotplot they generate:

```R
...
markers<- presto::wilcoxauc(pbmc, 'seurat_clusters', assay = 'data')
markers<- presto::top_markers(markers, n = 10, auc_min = 0.5, pct_in_min = 20, pct_out_max = 20)

markers
all_markers<- markers %>%
    select(-rank) %>% 
    unclass() %>% 
    stack() %>%
    pull(values) %>%
    unique() %>%
    .[!is.na(.)]

# Seurats dot plot
p<- DotPlot(object = pbmc, features = all_markers)
p
```
image

Whereas this is what we get without filtering by `pct.in` and `pct.out`:

```R
markers<- presto::wilcoxauc(pbmc, 'seurat_clusters', assay = 'data')
markers<- presto::top_markers(markers, n = 10)

markers
all_markers<- markers %>%
    select(-rank) %>% 
    unclass() %>% 
    stack() %>%
    pull(values) %>%
    unique() %>%
    .[!is.na(.)]

# Seurats dot plot
p<- DotPlot(object = pbmc, features = all_markers)
p
```
image

This will also help the heatmap markers

File: r/R/dotplot.R [hunk 13]
```diff
@@ -1,35 +1,84 @@
...
+#'
+#' @return
+#' @export
+#'
+#' @examples
 runDotPlot <- function(req, data) {
-  type <- req$body$type
-  req_genes <- req$body$genes
+  useMarkerGenes <- req$body$useMarkerGenes
+  data$dotplot.groups <- NA
```
Comment (alexvpickering) [inline r/R/dotplot.R]: maybe put this down where it gets set

File: r/R/dotplot.R [hunk 14]
```diff
@@ -1,35 +1,84 @@
...
+#' @param data
+#'
+#' @return
+#' @export
+#'
+#' @examples
 runDotPlot <- function(req, data) {
-  type <- req$body$type
-  req_genes <- req$body$genes
+  useMarkerGenes <- req$body$useMarkerGenes
```
Comment (alexvpickering) [inline r/R/dotplot.R]:
```suggestion
  use_marker_genes <- req$body$useMarkerGenes
```

check elsewhere that not using camelCase except for function names as previously discussed

File: r/R/dotplot.R [hunk 15]
```diff
@@ -1,35 +1,84 @@
...
-    filtered_cells <- intersect(set$cellIds, cells_id)
-    data$custom[cells_id %in% filtered_cells] <- i
+  #Collect ids to subset object
+  if (apply_filter) {
+    subsetIds <- filter_by$cellIds
+  } else {
+    subsetIds <- list()
+    for (child in filter_by$children) {
+      subsetIds <- append(subsetIds, child$cellIds)
+    }
```
Comment (alexvpickering) [inline r/R/dotplot.R]:
```suggestion
    subset_ids <- unlist(lapply(filter_by$children, `[[`, 'cellIds'))
```

untested, could break into two lines if you use. Not sure why your creating a list of lists of cell ids instead of a vector?

File: r/R/dotplot.R [hunk 16]
```diff
@@ -1,35 +1,84 @@
...
   }
 
-  if (type == "custom") {
-    annot <- data@misc$gene_annotations
-    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
-    annot_subset <- annot_subset[, c("input", "name")]
+  #Subset cell Ids
+  if (length(subsetIds)) {
+    data <- subset_ids(data,subsetIds)
```
Comment (alexvpickering) [inline r/R/dotplot.R]:
```suggestion
    data <- subset_ids(data, subsetIds)
```

run `styler`

File: r/R/dotplot.R [hunk 17]
```diff
@@ -1,35 +1,84 @@
...
+  for (i in seq_along(group_by_cell_sets)) {
+    cell_set <- group_by_cell_sets[[i]]
+    filtered_cells <- intersect(cell_set$cellIds, cells_id)
+    #This covers a border case where two cell_sets have the same name (but different ID). Can happen in scratchpad
+    if (cell_set$name %in% data$dotplot.groups) {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- paste0(cell_set$name, i)
+    } else {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- cell_set$name
+    }
+  }
```
Comment (alexvpickering) [inline r/R/dotplot.R]: maybe just pull out the `cell_set$name`'s first and uniqueify. Something like this (untested):

```R
cell_set_names <- make.unique(sapply(group_by_cell_sets, `[[`, 'name'))
for (i in seq_along(group_by_cell_sets)) {
  cell_set <- group_by_cell_sets[[i]]
  cell_set_name <- cell_set_names[i]
    ...
}
```

File: r/R/dotplot.R [hunk 18]
```diff
@@ -1,35 +1,84 @@
...
+  # If NA values are left in the group, dotplot function will fail.
+  data <- subset(data, subset = dotplot.groups != "NA")
```
Comment (alexvpickering) [inline r/R/dotplot.R]:
```suggestion
  data <- subset(data, subset = !is.na(dotplot.groups))
```

untested, but wouldn't typically check for `NA` by testing for equality with the string `"NA"` as it would break if the user defines a dotplot group as "NA"?

**Review by alexvpickering — DISMISSED** [review]
See comments so far. Haven't finished reviewing yet but happy to continue next week.

File: r/R/dotplot.R (hunk 11 above)
Comment (ogibson) [reply r/R/dotplot.R]: I had originally changed to avg.exp.scaled but switched back

File: r/R/dotplot.R (hunk 18 above)
Comment (ogibson) [reply r/R/dotplot.R]: Oh this was sloppy on my part, you are right

File: r/R/dotplot.R (hunk 15 above)
Comment (ogibson) [reply r/R/dotplot.R]: It's not a list of lists, the append function will add the elements of each cellIds list at the end of the subsetIds, returning a one level list of cells Id. That being said, I like your solution more :D

File: r/R/dotplot.R [hunk 19]
```diff
@@ -1,35 +1,81 @@
...
+    cells_id <- data$cells_id
+  } else {
+    message("The ids to subset the object are empty. Returning empty results.")
+    return(list())
+  }
 
-  dotplot_data <- dotplot_data[, c("avg.exp", "pct.exp", "name", "id")]
-  colnames(dotplot_data) <- c("avgExpression", "cellsPercentage", "geneName", "cellSet")
+  #Construct the dotplot.groups slot
+  data$dotplot.groups <- NA
```
Comment (ivababukova) [inline r/R/dotplot.R]: change the . to a dash

**Review by ivababukova — APPROVED** [review]
